File: ArucoDection/src/generate_board.cc
#include <opencv2/aruco/charuco.hpp>
#include <opencv2/aruco.hpp>
#include <opencv2/opencv.hpp>
#include <opencv2/core.hpp>

namespace {
    const char *about = "Create a(n) ArUco/ChArUco board";

    const char* keys =
        "{@outfile  |<none> | Output Image}"
        "{w         |       | Number of squares in X direction}"
        "{h         |       | Number of squares in Y direction}"
        "{sl        |       | Square side length (in pixels)}"
        "{ml        |       | Marker side length (in pixels)}"
        "{d         |       | dictionary: DICT_4X4_50=0, DICT_4X4_100=1, DICT_4X4_250=2,"
        "DICT_4X4_100=3, DICT_5X5_50=4, DICT_5X5_100=5, DICT_5X5_250=6, DICT_5X5_1000=7"
        "DICT_6X6_50=8, DICT_6X6_100=9, DICT_6X6_250=10, DICT_6X6_1000=11, DICT_7X7_50=12,"
        "DICT_7X7_100=13, DICT_7X7_250=14, DICT_7X7_1000=15, DICT_ARUCO_ORIGINAL=16}"
        "{cd        |       | Input file with custom dictionary}"
        "{m         |       | Margin size (in pixels). Default is (squareLength-markerLength)}"
        "{ch        |false  | Generate a ChArUco board instead of ArUco}"
        "{bb        | 1     | Number of bits in marker borders}"
        "{si        |false  | show generated image}";
}



void drawBoard(cv::Ptr<cv::aruco::CharucoBoard> board, cv::Size imgSize, int margins, int borderBits)
{
    cv::Mat boardImage;
    std::cout << imgSize.area() << std::endl;
    board->draw(imgSize, boardImage, margins, borderBits);

    // Display image
    cv::namedWindow("Display Grid", cv::WINDOW_AUTOSIZE);
    cv::imshow("Display Grid", boardImage);
    // cv::imwrite("../imgs/grid.jpg", boardImage);

    cv::waitKey(0);
}

void drawBoard(cv::Ptr<cv::aruco::GridBoard> board, cv::Size imgSize, int margins, int borderBits)
{
    cv::Mat boardImage;
    board->draw(imgSize, boardImage, margins, borderBits);

    // Display image
    cv::namedWindow("Display Grid", cv::WINDOW_AUTOSIZE);
    cv::imshow("Display Grid", boardImage);
    // cv::imwrite("../imgs/grid.jpg", boardImage);

    cv::waitKey(0);
}

int main(int argc, char **argv)
{

    // Command line input (see keys above)
    cv::CommandLineParser parser(argc, argv, keys);
    parser.about(about);

    if (argc < 7)
    {
        parser.printMessage();
        return 0;
    }

    // Get input from command line
    int squaresX = parser.get<int>("w");
    int squaresY = parser.get<int>("h");
    int squareLength = parser.get<int>("sl");
    int markerLength = parser.get<int>("ml");
    int margins = squareLength- markerLength;

    bool genChArUco = parser.get<bool>("ch");

    if (parser.has("m"))
    {
        margins = parser.get<int>("m");
    }

    int borderBits = parser.get<int>("bb");
    int showImage = parser.get<bool>("si");

    std::string out = parser.get<std::string>(0);

    if (!parser.check())
    {
        parser.printErrors();
        return 0;
    }

    // Create a dictionary
    cv::Ptr<cv::aruco::Dictionary> dictionary
        = cv::aruco::getPredefinedDictionary(0);

    if (parser.has("d"))
    {
        auto dictionaryId = parser.get<int>("d");
        dictionary = cv::aruco::getPredefinedDictionary(cv::aruco::PREDEFINED_DICTIONARY_NAME(dictionaryId));
    }
    // TODO : The cv::aruco::Dictionary::readDictionary() function does not exist.
    //          Need to find an alternative to reading custom dictionaries
    // else if (parser.has("cd"))
    // {
    //     cv::FileStorage fs(parser.get<std::string>("cd"), cv::FileStorage::READ);
    //     // bool ok = dictionary->cv::aruco::Dictionary::readDictionary(fs.root()); // Doesn't exist
    // }
    else
    {
        std::cerr << "Dictionary not specified" << std::endl;
        return 0;
    }

    // Formulate image size
    cv::Size  imageSize;
    imageSize.width = squaresX * squareLength + 2 * margins;
    imageSize.height = squaresY * squareLength + 2 * margins;
    std::cout << "Width: " << squaresX * squareLength + 2 * margins << std::endl;
    std::cout << "Height: " << squaresY * squareLength + 2 * margins << std::endl;
    // std::cout << "Square Length: " << squareLength << std::endl;
    std::cout << "Image size: " << imageSize.area() << std::endl;

    // Base class for all boards (GridBoard, ChArUco, ...)
    cv::Ptr<cv::aruco::Board> board;

    if (genChArUco)
    {
        board = cv::aruco::CharucoBoard::create(squaresX, squaresY,
                                                squareLength,
                                                markerLength,
                                                dictionary);
        drawBoard(board.staticCast<cv::aruco::CharucoBoard>(), imageSize, margins, borderBits);
    }
    else
    {
        board = cv::aruco::GridBoard::create(5,7,0.04, 0.01, dictionary);
        drawBoard(board.staticCast<cv::aruco::GridBoard>(), imageSize, margins, borderBits);
    }


    return 0;
}

File: ArucoDection/scripts/GenerateBoard.py
from copy import deepcopy
import argparse
import hashlib
import os
import cv2
import json


# construct the argument parser and parse the arguments
ap = argparse.ArgumentParser()
ap.add_argument("-w", "--width", type=int, required=True,
    help="Numbers of horizontal squares in ChArUco Board")
ap.add_argument("-l", "--length", type=int, required=True,
    help="Numbers of vertical squares in ChArUco Board")
ap.add_argument("-sl", "--square-length", type=float, required=True,
    help="Square side length in pixels.")
ap.add_argument("-ml", "--marker-length", type=float, required=True,
    help="Marker length in pixels")

ap.add_argument("-d", "--dictionary", type=int, required=True,
    help="Dictionary: "
        "DICT_4X4_50=0, DICT_4X4_100=1, DICT_4X4_250=2,"
        "DICT_4X4_100=3, DICT_5X5_50=4, DICT_5X5_100=5, DICT_5X5_250=6, DICT_5X5_1000=7"
        "DICT_6X6_50=8, DICT_6X6_100=9, DICT_6X6_250=10, DICT_6X6_1000=11, DICT_7X7_50=12,"
        "DICT_7X7_100=13, DICT_7X7_250=14, DICT_7X7_1000=15, DICT_ARUCO_ORIGINAL=16"
        )

## Turn the command line agruments into a dictionary
args = vars(ap.parse_args())
image_json = deepcopy(args)

## Generate the ChAruco board
gridboard = cv2.aruco.CharucoBoard_create(
    squaresX=args.get('width'),
    squaresY=args.get('length'),
    squareLength=args.get('square_length'), # 0.04
    markerLength=args.get('marker_length'), # 0.02
    dictionary=cv2.aruco.Dictionary_get(args.get('dictionary'))
)
img = gridboard.draw(outSize=(988, 1400))

filename = 'images.json'

if not os.path.exists(os.path.join(os.getcwd(), 'images')):
    os.mkdir('images')

## Write the png and the JSON file
cv2.imwrite('images/temp.png', img)
image_json['hash'] = hashlib.md5(open('images/temp.png', 'rb').read()).hexdigest()
os.rename('images/temp.png', 'images/{}.png'.format(image_json['hash']))

## Write the meta data about the image to filename
if not os.path.exists(os.path.join(os.getcwd(), filename)):
    f = open(filename, 'w+')
    temp = [image_json]
    f.write(json.dumps(temp))
    f.close()
else:
    f = open(filename, 'r')
    data = json.loads(f.read())
    f.close()

    duplicate = False
    for img in data:
        if img['hash'] == image_json['hash']:
            img = image_json
            duplicate = True

    if not duplicate:
        data.append(image_json)

    f = open(filename, 'w')
    f.write(json.dumps(data))
    f.close()